File: ffassistant/api/wl.py
import datetime

from flask import Blueprint, abort, jsonify, request

from ffassistant.api import get_db

wl_bp = Blueprint("wl", __name__, url_prefix="/api/wl")

# Fantasy regular seasons have run 16-18 weeks across 2013-2025 — render an
# open-ended grid up to the max so the current season always has editable
# blank slots to fill in as it goes, matching the Grid tab's convention.
MAX_WEEK = 18
CLOSE_GAME_MARGIN = 6
PLAYOFF_ROUNDS = ("semifinal", "final", "third_place")


def _current_year() -> int:
    return datetime.date.today().year


@wl_bp.get("/league_history")
def get_league_history():
    db = get_db()
    rows = db.execute("SELECT league_history_id, name, active FROM league_history ORDER BY name").fetchall()
    return jsonify([dict(r) for r in rows])


@wl_bp.get("/games")
def get_games():
    """One card per league for the year — active leagues always get a card
    (even with zero rows yet, so this season's results have somewhere to go),
    inactive/historical leagues only show up for years they actually played.
    """
    db = get_db()
    year = request.args.get("year", type=int) or _current_year()

    leagues = db.execute(
        """
        SELECT DISTINCT lh.league_history_id, lh.name
        FROM league_history lh
        LEFT JOIN matchups m ON m.league_history_id = lh.league_history_id AND m.season = ?
        LEFT JOIN league_seasons ls ON ls.league_history_id = lh.league_history_id AND ls.season = ?
        WHERE lh.active = 1 OR m.matchup_id IS NOT NULL OR ls.league_season_id IS NOT NULL
        ORDER BY lh.name
        """,
        (year, year),
    ).fetchall()

    result = []
    for league in leagues:
        week_rows = db.execute(
            """
            SELECT week, outcome, points_for, points_against, playoff_round
            FROM matchups WHERE league_history_id = ? AND season = ?
            """,
            (league["league_history_id"], year),
        ).fetchall()
        by_week = {r["week"]: dict(r) for r in week_rows}

        weeks = []
        for wk in range(1, MAX_WEEK + 1):
            row = by_week.get(wk)
            if row is None:
                weeks.append(
                    {"week": wk, "outcome": None, "points_for": None, "points_against": None,
                     "differential": None, "playoff_round": None}
                )
            else:
                weeks.append({**row, "differential": row["points_for"] - row["points_against"]})

        totals = db.execute(
            """
            SELECT
                COALESCE(SUM(CASE WHEN outcome = 'W' THEN 1 ELSE 0 END), 0) AS wins,
                COALESCE(SUM(CASE WHEN outcome = 'L' THEN 1 ELSE 0 END), 0) AS losses,
                COALESCE(SUM(CASE WHEN outcome = 'T' THEN 1 ELSE 0 END), 0) AS ties,
                COALESCE(SUM(points_for), 0) AS points_for,
                COALESCE(SUM(points_against), 0) AS points_against
            FROM matchups WHERE league_history_id = ? AND season = ?
            """,
            (league["league_history_id"], year),
        ).fetchone()

        result.append(
            {
                "league_history_id": league["league_history_id"],
                "name": league["name"],
                "weeks": weeks,
                "totals": dict(totals),
            }
        )

    return jsonify(result)


@wl_bp.get("/weekly")
def get_weekly():
    """Net games-above-even per week across every league combined, plus a
    running cumulative total through the season."""
    db = get_db()
    year = request.args.get("year", type=int) or _current_year()

    rows = db.execute(
        """
        SELECT week,
               SUM(CASE WHEN outcome = 'W' THEN 1 ELSE 0 END) AS wins,
               SUM(CASE WHEN outcome = 'L' THEN 1 ELSE 0 END) AS losses,
               SUM(CASE WHEN outcome = 'T' THEN 1 ELSE 0 END) AS ties,
               SUM(points_for) AS points_for,
               SUM(points_against) AS points_against
        FROM matchups WHERE season = ?
        GROUP BY week ORDER BY week
        """,
        (year,),
    ).fetchall()

    weeks = []
    cumulative = 0
    for row in rows:
        net = row["wins"] - row["losses"]
        cumulative += net
        weeks.append(
            {
                "week": row["week"],
                "wins": row["wins"],
                "losses": row["losses"],
                "ties": row["ties"],
                "points_for": row["points_for"],
                "points_against": row["points_against"],
                "net_games_above_even": net,
                "cumulative_net": cumulative,
            }
        )

    return jsonify(weeks)


@wl_bp.get("/leagues")
def get_leagues_view():
    db = get_db()
    year = request.args.get("year", type=int) or _current_year()

    rows = db.execute(
        """
        SELECT lh.league_history_id, lh.name, ls.wins, ls.losses, ls.ties,
               ls.buy_in, ls.max_payout, ls.actual_payout, ls.finish_position
        FROM league_seasons ls
        JOIN league_history lh ON lh.league_history_id = ls.league_history_id
        WHERE ls.season = ?
        ORDER BY lh.name
        """,
        (year,),
    ).fetchall()

    return jsonify([dict(r) for r in rows])


@wl_bp.get("/close_games")
def get_close_games():
    """Fixed threshold (< 6-point margin), not relative/percentage — Keith's
    own number from how he used the legacy sheet."""
    db = get_db()
    year = request.args.get("year", type=int) or _current_year()

    rows = db.execute(
        """
        SELECT lh.name AS league_name, m.week, m.outcome,
               ABS(m.points_for - m.points_against) AS margin
        FROM matchups m
        JOIN league_history lh ON lh.league_history_id = m.league_history_id
        WHERE m.season = ? AND ABS(m.points_for - m.points_against) < ?
        ORDER BY margin ASC
        """,
        (year, CLOSE_GAME_MARGIN),
    ).fetchall()

    return jsonify([dict(r) for r in rows])


@wl_bp.get("/all_time")
def get_all_time():
    """PF/PA are the real gap-fill the legacy 'All Years' sheet never
    populated — summed here from matchups since league_seasons doesn't carry
    them. Win/loss/tie totals, years played, and finish counts come from
    league_seasons (the season-end figures Keith already tracked)."""
    db = get_db()

    season_rows = db.execute(
        """
        SELECT lh.league_history_id, lh.name,
               COALESCE(SUM(ls.wins), 0) AS wins,
               COALESCE(SUM(ls.losses), 0) AS losses,
               COALESCE(SUM(ls.ties), 0) AS ties,
               COUNT(ls.league_season_id) AS years_played,
               SUM(CASE WHEN ls.finish_position = 1 THEN 1 ELSE 0 END) AS firsts,
               SUM(CASE WHEN ls.finish_position = 2 THEN 1 ELSE 0 END) AS seconds,
               SUM(CASE WHEN ls.finish_position = 3 THEN 1 ELSE 0 END) AS thirds,
               COALESCE(SUM(ls.buy_in), 0) AS total_buy_in,
               COALESCE(SUM(ls.actual_payout), 0) AS total_actual_payout
        FROM league_history lh
        LEFT JOIN league_seasons ls ON ls.league_history_id = lh.league_history_id
        GROUP BY lh.league_history_id, lh.name
        ORDER BY lh.name
        """
    ).fetchall()

    pf_pa_rows = db.execute(
        """
        SELECT league_history_id, SUM(points_for) AS points_for, SUM(points_against) AS points_against
        FROM matchups GROUP BY league_history_id
        """
    ).fetchall()
    pf_pa_by_id = {r["league_history_id"]: r for r in pf_pa_rows}

    # Which specific seasons landed each finish, for hover detail on the
    # 1st/2nd/3rd counts — a bare count doesn't say when.
    finish_rows = db.execute(
        """
        SELECT league_history_id, season, finish_position
        FROM league_seasons
        WHERE finish_position IN (1, 2, 3)
        ORDER BY season
        """
    ).fetchall()
    finish_years_by_id = {}
    for r in finish_rows:
        by_position = finish_years_by_id.setdefault(r["league_history_id"], {1: [], 2: [], 3: []})
        by_position[r["finish_position"]].append(r["season"])

    result = []
    for row in season_rows:
        total_games = row["wins"] + row["losses"] + row["ties"]
        pf_pa = pf_pa_by_id.get(row["league_history_id"])
        finish_years = finish_years_by_id.get(row["league_history_id"], {1: [], 2: [], 3: []})
        result.append(
            {
                "league_history_id": row["league_history_id"],
                "name": row["name"],
                "wins": row["wins"],
                "losses": row["losses"],
                "ties": row["ties"],
                "win_pct": row["wins"] / total_games if total_games else None,
                "years_played": row["years_played"],
                "firsts": row["firsts"],
                "seconds": row["seconds"],
                "thirds": row["thirds"],
                "first_years": finish_years[1],
                "second_years": finish_years[2],
                "third_years": finish_years[3],
                "points_for": pf_pa["points_for"] if pf_pa else None,
                "points_against": pf_pa["points_against"] if pf_pa else None,
                "total_buy_in": row["total_buy_in"],
                "total_actual_payout": row["total_actual_payout"],
            }
        )

    return jsonify(result)


@wl_bp.get("/finishes")
def get_finishes():
    """League x year finish-position grid for the All Years tab's history
    table. Years span every season on record through the current year (so an
    active league's in-progress season still gets a column). A league with no
    league_seasons row for a given year (didn't exist yet, or a gap) gets
    `null` in that cell — same as a row that exists but has no recorded
    finish_position — since the client renders both as "no data" alike.
    """
    db = get_db()

    leagues = db.execute("SELECT league_history_id, name FROM league_history ORDER BY name").fetchall()

    season_bounds = db.execute("SELECT MIN(season) AS mn, MAX(season) AS mx FROM league_seasons").fetchone()
    min_year = season_bounds["mn"]
    max_year = max(season_bounds["mx"] or 0, _current_year()) if min_year else _current_year()
    years = list(range(min_year, max_year + 1)) if min_year else []

    finish_rows = db.execute("SELECT league_history_id, season, finish_position FROM league_seasons").fetchall()
    finish_by_league = {}
    for r in finish_rows:
        finish_by_league.setdefault(r["league_history_id"], {})[r["season"]] = r["finish_position"]

    result_leagues = []
    for league in leagues:
        by_year = finish_by_league.get(league["league_history_id"], {})
        result_leagues.append(
            {
                "league_history_id": league["league_history_id"],
                "name": league["name"],
                "finishes": {str(year): by_year.get(year) for year in years},
            }
        )

    return jsonify({"years": years, "leagues": result_leagues})


@wl_bp.put("/matchups")
def put_matchup():
    """Manual entry/edit for a single week's result — the baseline workflow
    (Keith has always filled these in by hand); outcome is derived here from
    the two scores rather than entered separately."""
    db = get_db()
    body = request.get_json(force=True)

    league_history_id = body.get("league_history_id")
    season = body.get("season")
    week = body.get("week")
    points_for = body.get("points_for")
    points_against = body.get("points_against")
    playoff_round = body.get("playoff_round")

    if not league_history_id or not season or not week or points_for is None or points_against is None:
        abort(400, description="league_history_id, season, week, points_for, and points_against are required")
    if playoff_round is not None and playoff_round not in PLAYOFF_ROUNDS:
        abort(400, description=f"playoff_round must be one of {PLAYOFF_ROUNDS} or null")

    if points_for > points_against:
        outcome = "W"
    elif points_for < points_against:
        outcome = "L"
    else:
        outcome = "T"

    db.execute(
        """
        INSERT INTO matchups (league_history_id, season, week, points_for, points_against, outcome, playoff_round)
        VALUES (?, ?, ?, ?, ?, ?, ?)
        ON CONFLICT (league_history_id, season, week) DO UPDATE SET
            points_for = excluded.points_for,
            points_against = excluded.points_against,
            outcome = excluded.outcome,
            playoff_round = excluded.playoff_round
        """,
        (league_history_id, season, week, points_for, points_against, outcome, playoff_round),
    )
    db.commit()

    return jsonify(
        {
            "league_history_id": league_history_id,
            "season": season,
            "week": week,
            "points_for": points_for,
            "points_against": points_against,
            "outcome": outcome,
            "playoff_round": playoff_round,
        }
    )

File: static/js/wl.js
import { api } from "./api.js";

const PLAYOFF_LABELS = {
  semifinal: "Semifinal",
  final: "Final",
  third_place: "3rd Place",
};

export async function renderWlView(container, state) {
  const view = state.wlTab; // "games" | "weekly" | "leagues" | "close_games" | "all_time"
  const wrap = el("div", "wl-view");

  if (view === "games") {
    wrap.appendChild(await buildGamesView(state.wlYear, () => renderWlView(container, state)));
  } else if (view === "weekly") {
    wrap.appendChild(await buildWeeklyView(state.wlYear));
  } else if (view === "leagues") {
    wrap.appendChild(await buildLeaguesView(state.wlYear));
  } else if (view === "close_games") {
    wrap.appendChild(await buildCloseGamesView(state.wlYear));
  } else if (view === "all_time") {
    wrap.appendChild(await buildAllTimeView());
  }

  container.appendChild(wrap);
}

// ---- Games: one card per league, weeks as rows, click a week to edit ----

async function buildGamesView(year, refresh) {
  const leagues = await api.getWlGames(year);

  const scrollWrap = el("div", "wl-games-scroll");
  if (leagues.length === 0) {
    const empty = document.createElement("p");
    empty.className = "wl-empty";
    empty.textContent = "No leagues found for this year.";
    scrollWrap.appendChild(empty);
    return scrollWrap;
  }

  for (const league of leagues) {
    scrollWrap.appendChild(buildLeagueGameCard(league, year, refresh));
  }
  return scrollWrap;
}

function buildLeagueGameCard(league, year, refresh) {
  const card = el("div", "wl-game-card");

  const heading = document.createElement("h3");
  heading.textContent = league.name;
  card.appendChild(heading);

  const table = el("table", "wl-games-table");
  const thead = document.createElement("thead");
  thead.innerHTML = "<tr><th>Wk</th><th>Outcome</th><th>PF</th><th>PA</th><th>Diff</th><th>Playoffs</th></tr>";
  table.appendChild(thead);

  const tbody = document.createElement("tbody");
  for (const weekData of league.weeks) {
    tbody.appendChild(buildWeekRow(league.league_history_id, weekData, year, refresh));
  }
  table.appendChild(tbody);

  const tfoot = document.createElement("tfoot");
  const totals = league.totals;
  const totalRow = document.createElement("tr");
  totalRow.innerHTML =
    `<td>Total</td><td>${totals.wins}-${totals.losses}-${totals.ties}</td>` +
    `<td>${totals.points_for.toFixed(1)}</td><td>${totals.points_against.toFixed(1)}</td>` +
    `<td colspan="2"></td>`;
  tfoot.appendChild(totalRow);
  table.appendChild(tfoot);

  card.appendChild(table);
  return card;
}

function buildWeekRow(leagueHistoryId, weekData, year, refresh) {
  const row = document.createElement("tr");
  if (weekData.outcome == null) row.classList.add("wl-week-empty");

  renderWeekRowContent(row, weekData);

  row.addEventListener("click", () => {
    if (row.querySelector(".wl-edit-form")) return;
    openWeekEditForm(row, leagueHistoryId, weekData, year, refresh);
  });

  return row;
}

function renderWeekRowContent(row, weekData) {
  row.innerHTML = "";

  const outcomeCell = td(weekData.outcome || "—");
  if (weekData.outcome) outcomeCell.classList.add(`wl-outcome-${weekData.outcome}`);

  const diffCell = td(weekData.differential != null ? weekData.differential.toFixed(1) : "—");
  applyDiffClass(diffCell, weekData.differential);

  row.append(
    td(weekData.week),
    outcomeCell,
    td(weekData.points_for != null ? weekData.points_for.toFixed(1) : "—"),
    td(weekData.points_against != null ? weekData.points_against.toFixed(1) : "—"),
    diffCell,
    td(PLAYOFF_LABELS[weekData.playoff_round] || "")
  );
}

function applyDiffClass(cell, value) {
  if (value == null || value === 0) return;
  cell.classList.add(value > 0 ? "wl-diff-positive" : "wl-diff-negative");
}

function openWeekEditForm(row, leagueHistoryId, weekData, year, refresh) {
  row.innerHTML = "";
  const cell = document.createElement("td");
  cell.colSpan = 6;
  cell.className = "wl-edit-form";

  const pfInput = document.createElement("input");
  pfInput.type = "number";
  pfInput.step = "0.01";
  pfInput.placeholder = "PF";
  pfInput.value = weekData.points_for ?? "";

  const paInput = document.createElement("input");
  paInput.type = "number";
  paInput.step = "0.01";
  paInput.placeholder = "PA";
  paInput.value = weekData.points_against ?? "";

  const playoffSelect = document.createElement("select");
  const options = { "": "—", semifinal: "Semifinal", final: "Final", third_place: "3rd Place" };
  for (const [value, label] of Object.entries(options)) {
    const opt = document.createElement("option");
    opt.value = value;
    opt.textContent = label;
    if ((weekData.playoff_round || "") === value) opt.selected = true;
    playoffSelect.appendChild(opt);
  }

  const saveBtn = document.createElement("button");
  saveBtn.type = "button";
  saveBtn.className = "wl-save-button";
  saveBtn.textContent = "Save";
  saveBtn.addEventListener("click", async () => {
    if (pfInput.value === "" || paInput.value === "") return;
    await api.putWlMatchup({
      league_history_id: leagueHistoryId,
      season: year,
      week: weekData.week,
      points_for: Number(pfInput.value),
      points_against: Number(paInput.value),
      playoff_round: playoffSelect.value || null,
    });
    refresh();
  });

  cell.append(pfInput, paInput, playoffSelect, saveBtn);
  row.appendChild(cell);
  pfInput.focus();
}

// ---- Weekly: net games-above-even across all leagues, plus cumulative ----

async function buildWeeklyView(year) {
  const weeks = await api.getWlWeekly(year);

  const table = el("table", "wl-data-table wl-weekly-table");
  const thead = document.createElement("thead");
  thead.innerHTML =
    "<tr><th>Week</th><th>W</th><th>L</th><th>T</th><th>PF</th><th>PA</th><th>Diff</th><th>Net</th><th>Cumulative</th></tr>";
  table.appendChild(thead);

  const tbody = document.createElement("tbody");
  for (const week of weeks) {
    const diff = week.points_for - week.points_against;
    const diffCell = td(diff.toFixed(1));
    applyDiffClass(diffCell, diff);

    const row = document.createElement("tr");
    row.append(
      td(week.week),
      td(week.wins),
      td(week.losses),
      td(week.ties),
      td(week.points_for.toFixed(1)),
      td(week.points_against.toFixed(1)),
      diffCell,
      td(formatSigned(week.net_games_above_even)),
      td(formatSigned(week.cumulative_net))
    );
    tbody.appendChild(row);
  }
  table.appendChild(tbody);
  return table;
}

// ---- Leagues: one row per league for the year ----

async function buildLeaguesView(year) {
  const leagues = await api.getWlLeagues(year);

  const table = el("table", "wl-data-table wl-leagues-table");
  const thead = document.createElement("thead");
  thead.innerHTML =
    "<tr><th>League</th><th>W</th><th>L</th><th>T</th><th>Buy-in</th><th>Max</th><th>Actual</th><th>Finish</th></tr>";
  table.appendChild(thead);

  const tbody = document.createElement("tbody");
  for (const league of leagues) {
    const row = document.createElement("tr");
    row.append(
      td(league.name),
      td(league.wins),
      td(league.losses),
      td(league.ties),
      td(formatMoney(league.buy_in)),
      td(formatMoney(league.max_payout)),
      td(formatMoney(league.actual_payout)),
      td(league.finish_position ?? "—")
    );
    tbody.appendChild(row);
  }
  table.appendChild(tbody);
  return table;
}

// ---- Close Games: fixed <6pt margin, split into wins vs. losses ----

async function buildCloseGamesView(year) {
  const games = await api.getWlCloseGames(year);
  const wrap = el("div", "wl-close-games-wrap");

  wrap.appendChild(buildCloseGamesTable("Close Wins", games.filter((g) => g.outcome === "W"), "wl-outcome-W"));
  wrap.appendChild(buildCloseGamesTable("Close Losses", games.filter((g) => g.outcome === "L"), "wl-outcome-L"));
  return wrap;
}

function buildCloseGamesTable(title, games, outcomeClass) {
  const section = el("div", "wl-close-games-section");
  const heading = document.createElement("h3");
  heading.textContent = title;
  section.appendChild(heading);

  const table = el("table", "wl-data-table wl-close-games-table");
  const thead = document.createElement("thead");
  thead.innerHTML = "<tr><th>League</th><th>Week</th><th>Margin</th></tr>";
  table.appendChild(thead);

  const tbody = document.createElement("tbody");
  if (games.length === 0) {
    const row = document.createElement("tr");
    const cell = td(`No ${title.toLowerCase()} within 6 points this year.`);
    cell.colSpan = 3;
    row.appendChild(cell);
    tbody.appendChild(row);
  }
  for (const game of games) {
    const row = document.createElement("tr");
    const marginCell = td(game.margin.toFixed(1));
    marginCell.classList.add(outcomeClass);
    row.append(td(game.league_name), td(game.week), marginCell);
    tbody.appendChild(row);
  }
  table.appendChild(tbody);
  section.appendChild(table);
  return section;
}

// ---- All Years: one rollup row per league, spans every season ----

async function buildAllTimeView() {
  const [leagues, finishes] = await Promise.all([api.getWlAllTime(), api.getWlFinishes()]);

  const wrap = el("div", "wl-all-time-wrap");

  const table = el("table", "wl-data-table wl-all-time-table");
  const thead = document.createElement("thead");
  thead.innerHTML =
    "<tr><th>League</th><th>W</th><th>L</th><th>T</th><th>Win %</th><th>Years</th>" +
    "<th>1st</th><th>2nd</th><th>3rd</th><th>PF</th><th>PA</th><th>Total Buy-in</th><th>Total Payout</th></tr>";
  table.appendChild(thead);

  const tbody = document.createElement("tbody");
  for (const league of leagues) {
    const row = document.createElement("tr");
    row.append(
      td(league.name),
      td(league.wins),
      td(league.losses),
      td(league.ties),
      td(league.win_pct != null ? (league.win_pct * 100).toFixed(1) + "%" : "—"),
      td(league.years_played),
      finishTd(league.firsts, league.first_years),
      finishTd(league.seconds, league.second_years),
      finishTd(league.thirds, league.third_years),
      td(league.points_for != null ? league.points_for.toFixed(1) : "—"),
      td(league.points_against != null ? league.points_against.toFixed(1) : "—"),
      td(formatMoney(league.total_buy_in)),
      td(formatMoney(league.total_actual_payout))
    );
    tbody.appendChild(row);
  }
  table.appendChild(tbody);
  wrap.appendChild(table);

  wrap.appendChild(buildFinishesGrid(finishes));
  return wrap;
}

// ---- Finishes grid: league x year, gold/silver/bronze for 1st/2nd/3rd ----

function buildFinishesGrid(finishes) {
  const scrollWrap = el("div", "wl-finishes-scroll");

  const heading = document.createElement("h3");
  heading.textContent = "Finishes by Year";
  scrollWrap.appendChild(heading);

  const table = el("table", "wl-data-table wl-finishes-table");
  const thead = document.createElement("thead");
  const headRow = document.createElement("tr");
  headRow.appendChild(document.createElement("th")).textContent = "League";
  for (const year of finishes.years) {
    const th = document.createElement("th");
    th.textContent = year;
    headRow.appendChild(th);
  }
  thead.appendChild(headRow);
  table.appendChild(thead);

  const tbody = document.createElement("tbody");
  for (const league of finishes.leagues) {
    const row = document.createElement("tr");
    row.appendChild(td(league.name));
    for (const year of finishes.years) {
      row.appendChild(finishCell(league.finishes[year]));
    }
    tbody.appendChild(row);
  }
  table.appendChild(tbody);
  scrollWrap.appendChild(table);

  return scrollWrap;
}

function finishCell(place) {
  const cell = document.createElement("td");
  if (place == null) {
    cell.textContent = "";
    cell.className = "wl-finish-none";
    return cell;
  }
  cell.textContent = place;
  cell.className =
    place === 1 ? "wl-finish-gold" : place === 2 ? "wl-finish-silver" : place === 3 ? "wl-finish-bronze" : "";
  return cell;
}

function formatSigned(n) {
  return n > 0 ? `+${n}` : `${n}`;
}

function formatMoney(n) {
  return n != null ? `$${n}` : "—";
}

function td(content) {
  const cell = document.createElement("td");
  cell.textContent = content;
  return cell;
}

// A finish count (1st/2nd/3rd) with the specific years on hover — the raw
// count alone doesn't say when.
function finishTd(count, years) {
  const cell = td(count);
  if (years && years.length) {
    cell.title = years.join(", ");
    cell.classList.add("wl-finish-cell");
  }
  return cell;
}

function el(tag, className) {
  const node = document.createElement(tag);
  if (className) node.className = className;
  return node;
}
